### DISCUSSION TITLE
optional structure fields

### DISCUSSION THREAD

**Discussion by wangjiawen2013** [tracel-ai/burn#2591] [Q&A] [answered]
Hi,
Here is a chunk of pytorch code that can generate a block with or without batchnorm:
![image](https://github.com/user-attachments/assets/8704f436-4411-41c4-a5c5-6ac35ddd69c0)

When I use burn instead, I find it's hard to implement similar function. How to set `norm` as an optional field ?
```rust
#[derive(Module, Debug)]
pub struct Block<B: Backend> {
    fc: nn::Linear<B>,
    norm: nn::BatchNorm<B, 1>,
    activation: nn::LeakyRelu,
}

impl<B: Backend> Block<B> {
    pub fn new(in_features: usize, out_features: usize, device: &B::Device) -> Self {
        let fc = nn::LinearConfig::new(in_features, out_features)
            .with_bias(True)
            .init(device);
        let norm = nn::BatchNormConfig::new(out_features, 0.8).init(device);
        let activation = nn::LeakyReluConfig::new().with_negative_slope(0.2).init(device);

        Self {
            fc,
            norm,
            activation
        }
    }
}
```

**Comment by laggui** [comment, marked as answer]
This can be achieved via a simple `Option` field.

Take this example from the [ResNet implementation](https://github.com/tracel-ai/models/blob/main/resnet-burn/resnet/src/block.rs#L59-L90):

```rust
#[derive(Module, Debug)]
pub struct BasicBlock<B: Backend> {
    conv1: Conv2d<B>,
    bn1: BatchNorm<B, 2>,
    relu: Relu,
    conv2: Conv2d<B>,
    bn2: BatchNorm<B, 2>,
    downsample: Option<Downsample<B>>, // optional
}

impl<B: Backend> BasicBlock<B> {
    fn forward(&self, input: Tensor<B, 4>) -> Tensor<B, 4> {
        let identity = input.clone();

        let out = self.conv1.forward(input);
        let out = self.bn1.forward(out);
        let out = self.relu.forward(out);
        let out = self.conv2.forward(out);
        let out = self.bn2.forward(out);

        let out = {
           // Optional module
            match &self.downsample {
                Some(downsample) => out + downsample.forward(identity),
                None => out + identity,
            }
        };

        // Activation
        self.relu.forward(out)
    }
}
```
